### ISSUE TITLE
Prefetch performance

### ISSUE THREAD

**Issue by DGolubets** [functional-streams-for-scala/fs2#1098]
Hi,
The following simple example of a stream, copying a file, has drastically different performance with and without prefetching.

```
    import scala.concurrent.ExecutionContext.Implicits.global
    io
      .file
      .readAll[IO](sourcePath, 4096)
      .prefetch
      .through(io.file.writeAll(destPath))
      .compile
      .drain
      .unsafeRunSync()
```

I will throw some numbers I got:

File size **150 KB**, **without** prefetch: **5ms**
File size **150 KB**, **with** prefetch: **18ms**

File size **500 MB**, **without** prefetch: **3s**
File size **500 MB**, **with** prefetch: **30s**

I tested on:
Scala 2.12.4
FS2 0.10.1

**Comment by mpilquist** [comment]
I suspect the bug described in #1107 is the root cause of this performance issue. Once we fix #1107, let's retest performance of prefetch.

**Comment by DGolubets** [comment]
It's the same with 0.10.3

**Comment by pchlupacek** [comment]
i wonder how prefetch with ´repeatEval(dequeue1))´ will perform

**Comment by DGolubets** [comment]
@pchlupacek , do you mean manual prefetching with a queue?
```scala
Stream
      .eval(fs2.async.boundedQueue[IO, Option[Chunk[Byte]]](1))
      .flatMap { q =>
        val prefetch = io.file
          .readAll[IO](sourcePath, 4096)
          .chunks
          .map(Some(_))
          .append(Stream.emit(None))
          .through(q.enqueue)
          .compile
          .drain

        val write = Stream
          .repeatEval(q.dequeue1)
          .takeWhile(_.isDefined)
          .map(_.get)
          .flatMap(chunk => Stream.chunk(chunk))
          .through(io.file.writeAll(destPath))
          .compile
          .drain

        Stream.eval(write).merge {
          Stream.eval(prefetch)
        }
      }
      .compile
      .drain
      .unsafeRunSync()
```
Having an explicit queue gives better performance than prefetch (only x1.5 - x2 times slower).

**Comment by SystemFw** [comment]
the weird thing here is that you have almost exactly replicated the implementation of `prefetch`
```scala
    /** Alias for `prefetchN(1)`. */
    def prefetch(implicit ec: ExecutionContext, F: Effect[F]): Stream[F, O] = prefetchN(1)

    /**
      * Behaves like `identity`, but starts fetches up to `n` segments in parallel with downstream
      * consumption, enabling processing on either side of the `prefetchN` to run in parallel.
      */
    def prefetchN(n: Int)(implicit ec: ExecutionContext, F: Effect[F]): Stream[F, O] =
      Stream.eval(async.boundedQueue[F, Option[Segment[O, Unit]]](n)).flatMap { queue =>
        queue.dequeue.unNoneTerminate
          .flatMap(Stream.segment(_))
          .concurrently(self.segments.noneTerminate.to(queue.enqueue))
      }
```

**Comment by DGolubets** [comment]
The biggest difference comes from the way read and rights streams are executed here I guess.
In my code I create 2 IO effects and then start their concurrent evaluation. If I remove  ```.compile.drain``` parts and feed two streams directly to ```concurrently``` it runs slow.

Another is ```Stream.repeatEval(q.dequeue1)``` vs ```q.dequeue``` as the latter clearly does more complicated stuff under the hood.

**Comment by DGolubets** [comment]
I traced it down to ```.interruptScope``` being the slow part. But I think that's how far I wish to go - ```Algebra``` looks too complicated to dive to. I will see later what is the slow part of ```dequeue``` though.

**Comment by pchlupacek** [comment]
@DGolubets this shall be addressed now with #1258, that is in 1.0 master branch. I wonder if you can check if that is improving your use-case?

**Comment by DGolubets** [comment]
@pchlupacek It is! The overhead of prefetch is much lower now (just about 20% slower).
